### PR TITLE
MLX-LM batching 

### PR DESCRIPTION
**Pull request by shepardxia**
Adding batching function to MLX-LM backend with KV caching.

### CODE REVIEW

**Comment by shepardxia** [comment]
@benlebrun PR ready for review!

File: genlm/backend/llm/mlx.py [hunk 1]
```diff
@@ -139,64 +240,141 @@ def _step(input_tokens: mx.array):
...
+
+            query_groups = defaultdict(list)
+            for query in queries:
+                key = tuple(query.prompt)
+                query_groups[key].append(query)
+
+            # Use one representative query from each group
+            unique_queries = [group[0] for group in query_groups.values()]
+
+            input_prompts = [q.prompt for q in unique_queries]
```
Comment (benlebrun) [inline genlm/backend/llm/mlx.py]: should be inside the if statement if it is only used if self.batching is true

File: genlm/backend/llm/mlx.py [hunk 2]
```diff
@@ -39,17 +48,102 @@ def from_name(cls, *args, **kwargs):  # pragma: no cover
 
 else:
 
+    def _to_torch(logprobs):
```
Comment (benlebrun) [inline genlm/backend/llm/mlx.py]: should these be static methods in the class?

File: tests/test_mlx.py [hunk 3]
```diff
@@ -2,19 +2,27 @@
 import asyncio
 import torch
 from arsenal.maths import compare
-from genlm.backend.llm import load_model_by_name
+from genlm.backend.llm import load_model_by_name, AsyncMlxLM
 
+TOLERANCES = {
+    "yujiepan/mamba2-tiny-random": 1.5,
```
Comment (benlebrun) [inline tests/test_mlx.py]: This is a very high tolerance. what's the difference between these two models?

**Review by benlebrun — CHANGES_REQUESTED** [review]
Looks good, left a few comments---I think this can be cleaned up a bit, but generally looks good.

One thing to consider: we don't actually need the timeout functionality for async batching. The basic idea is to have a queue and a background task that eagerly pulls from the queue and processes all requests together in a batch. 

Because of the way that the `asyncio` scheduler works, all concurrent requests will be batched together. The background task grabs one item with an await, and then drains the queue with get_nowait() to form the batch. Because other coroutines that call _queue_request will run while the background loop awaits the first get(), those concurrent requests will land in the same batch.

A rough sketch looks like this:

```python
class AutoBatchedSketch:

    def __init__(self):
        self._queue = None
        self._task = None
    
    def _start(self):
        if not self._task or self._task.done():
            self._queue = asyncio.Queue()
            self._task = asyncio.create_task(self._background_loop())

    def _queue_request(self, request):
        if not self._task or self._task.done():
            self._start()

        future = asyncio.get_running_loop().create_future()
        self._queue.put_nowait((request, future))
        return future

    async def next_token_logprobs(self, token_ids) -> Any:
        """ Public API. Enqueue a request and await its result. """
        return await self._queue_request(token_ids)

    async def _background_loop(self):
        while True:
            try:
                requests = [await self._queue.get()]

                try:
                    while True: 
                        requests.append(self._queue.get_nowait())
                except asyncio.QueueEmpty:
                    pass

                    inputs, futures = zip(*requests)
                    results = self._batch_call(inputs)
                    for future, result in zip(futures, results):
                        future.set_result(result)

            except Exception as e:
                for _, future in requests:
                    if not future.done():
                        future.set_exception(e)
                raise
```

Not saying that we need to implement this approach now, but it is worth keeping in mind if you want a more efficient approach which doesn't require specifying a batch size and a timeout. We used it here https://github.com/genlm/genlm-bytes/blob/e76ca6908b2360690e5ecf098b377395b342978a/genlm/bytes/trie.py#L484.

File: genlm/backend/llm/mlx.py [hunk 4]
```diff
@@ -39,120 +47,397 @@ def from_name(cls, *args, **kwargs):  # pragma: no cover
...
-            self.cache = (
-                OutputMLXCache(maxsize=cache_size, **cache_opts)
-                if cache_size > 0
-                else None
-            )
+            self.cache = DynamicTokenTrie()
             self.generation_stream = mx.new_stream(mx.default_device())
-
+            self.queries = []
+            self.batch_size = (
```
Comment (benlebrun) [inline genlm/backend/llm/mlx.py]: add a warning to let the user know that the model is not batchable.

**Review by benlebrun — APPROVED** [review]
Looks great! Just left a few comments, then should be good to merge!
